### DISCUSSION TITLE
how to ignore "\r\n" at most time, but recognize in some specific condition

### DISCUSSION THREAD

**Discussion by ReverseJay** [antlr/antlr4#4008] [Q&A]
i am developing a simple cpp recognition program, at most time, i need to ignore "\r\n", but when i deal with "#define", i could only use "\r\n" as terminal
```
void aLongLongFunction(int var1, int var2, int var3,      //need to ignore "\r\n"
         int var4, int var5);
#define SomeAlternative(TypeA) \                          //need to ignore "\r\n"
         void aFunc##TypeA(){ \                           //need to ignore "\r\n"
                 //... \                                  //need to ignore "\r\n"
         }                                                //need to recognize "\r\n"
```

**Comment by ReverseJay** [comment]
now i use hidden channel to do some extra judging, is there a more efficient way to do

**Reply by kaby76** [reply]
The ISO Spec for C++ states specifically that preprocessing is performed in a separate step. You can write a preprocessor with Antlr, but it's not particularly easy because the Spec blurs the line between token definitions between preprocessing and post-preprocessing. Thus, there is a pull between sharing the grammars for pre- and post-processing vs. yanking them completely apart. This is a large task--I know because I implemented this.

If you want to recognize macro defs without actually expanding them, then push the entire recognition of the preprocessor directive into the lexer. Antlr lexer rules are context-free grammars, so everything you can do in the parser, can be done in the lexer. Multi-line preprocessing directives are recognized this way in the C++ grammar in grammars-v4. See https://github.com/antlr/grammars-v4/blob/8a0f7f737b06d507fc0ffb7e86f199aa7b6ed33f/cpp/CPP14Lexer.g4#L31.

**Reply by ReverseJay** [reply]
thanks, i'll try to write a preprocessor
